test/org/zaproxy/zap/extension/ascanrulesBeta/CsrftokenscanUnitTest.java

```diff
@@ -163,6 +167,48 @@ public void shouldProcessWithTwoSessionCookiesAndOtherCookie() throws Exception
         assertThat(httpMessagesSent, hasSize(greaterThanOrEqualTo(1)));
         assertThat(httpMessagesSent.get(0).getCookieParams(), hasSize(2)); // 2 session cookies
     }
+    
+    @Test
+    public void shouldNotProcessAtHighThresholdAndOutOfScope() throws HttpMalformedHeaderException, URIException {
+        // Given
+        HttpMessage msg = getAntiCSRFCompatibleMessage();
+        rule.setConfig(new ZapXmlConfiguration());
+        rule.setAlertThreshold(AlertThreshold.HIGH);
+        Model.getSingleton().getSession().getNewContext("TestContext").addIncludeInContextRegex(".*example.com.*");
+        rule.init(msg, parent);
+        // When
+        this.rule.scan();
+        // Then
+        assertThat(httpMessagesSent, hasSize(equalTo(0)));
+    }
+    
+    @Test
+    public void shouldProcessAtMediumThresholdAndOutOfScope() throws HttpMalformedHeaderException, URIException {
+        // Given
+        HttpMessage msg = getAntiCSRFCompatibleMessage();
+        rule.setConfig(new ZapXmlConfiguration());
+        rule.setAlertThreshold(AlertThreshold.MEDIUM);
+        // Note: This Test leverages the context setup in a previous test
```

Comment:
I looked at setting up an `@After` method in the unit test to clean this up between tests or wrapping each test in a try/finally block. But upon looping through the contexts and trying to delete context it would lead to an NPE.

Reply:
Are we ok with the dependency between test functions?

Reply:
I checked this but forgot to post a comment. We would have to use `Whitebox` to change some internal state of the `Model` (or the`Session`) to get it working.

How about creating the `HttpMessage` and override the `isInScope` method to return true/false as needed? That would make it simpler.

Reply:
Tried doing this:
```java
    HttpMessage msg = new HttpMessage() {
        @Override 
        public boolean isInScope() {
            return false;
        }
    };
    msg.getRequestHeader().setMethod("GET");
    msg.getRequestHeader().setURI(new URI("http", "localhost", "/", ""));
    msg.setRequestBody("<html><form><input type=\"hidden\" name=\"customAntiCSRF\" value=" + Math.random() + "></input></form></html>");
``` 

It NPEs at `this.rule.scan();`

```trace
java.lang.NullPointerException
	at org.parosproxy.paros.network.HttpMessage.isInScope(HttpMessage.java:872)
	at org.zaproxy.zap.extension.ascanrulesBeta.Csrftokenscan.scan(Csrftokenscan.java:169)
	at org.zaproxy.zap.extension.ascanrulesBeta.CsrftokenscanUnitTest.shouldNotProcessAtHighThresholdAndOutOfScope(CsrftokenscanUnitTest.java:191)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.mockito.internal.runners.JUnit45AndHigherRunnerImpl.run(JUnit45AndHigherRunnerImpl.java:37)
	at org.mockito.runners.MockitoJUnitRunner.run(MockitoJUnitRunner.java:62)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:539)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:761)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:461)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:207)
```

Reply:
Sorry, should have checked that. The message is being cloned, we need to override on the cloned message instead:
```Java
HttpMessage msg = new HttpMessage() {

    @Override
    public HttpMessage cloneRequest() {
        HttpMessage newMsg = new HttpMessage() {

            @Override
            public boolean isInScope() {
                return false;
            }
        };

        if (!this.getRequestHeader().isEmpty()) {
            try {
                newMsg.getRequestHeader().setMessage(this.getRequestHeader().toString());
            } catch (HttpMalformedHeaderException e) {
                throw new RuntimeException(e);
            }
            newMsg.setRequestBody(this.getRequestBody().getBytes());
        }
        return newMsg;
    }
};
msg.getRequestHeader().setMethod("GET");
msg.getRequestHeader().setURI(new URI("http", "localhost", "/", ""));
msg.setRequestBody("<html><form><input type=\"hidden\" name=\"customAntiCSRF\" value=" + Math.random() + "></input></form></html>");
```

Reply:
Thanks!

Reply:
Something is still off, leveraging that within `shouldNotProcessAtHighThresholdAndOutOfScope()` works (the test is successful). [I think think is a false success.]

Using it within `shouldProcessAtMediumThresholdAndOutOfScope` or the Low one fails, zero messages are sent.

Reply:
Yeah, the `msg` is not complete (header empty and missing response), we need to initialise that too:
```Java
HttpMessage compatMsg = getAntiCSRFCompatibleMessage();
msg.setRequestHeader(compatMsg.getRequestHeader());
msg.setRequestBody(compatMsg.getRequestBody());
msg.setResponseHeader(compatMsg.getResponseHeader());
msg.setResponseBody(compatMsg.getResponseBody());
return msg;
```

Reply:
Success, thanks!